page_decoder.go

```diff
@@ -0,0 +1,399 @@
+	i := 0
+	n, err := r.readByteArrays(func(b []byte) bool {
+		k := plain.ByteArrayLengthSize + len(b)
+		if k > (len(values) - i) {
+			return false
+		}
+		plain.PutByteArrayLength(values[i:], len(b))
+		copy(values[i+plain.ByteArrayLengthSize:], b)
+		i += k
+		return true
+	})
+	if i == 0 && err == nil {
+		err = io.ErrShortBuffer
+	}
+	return n, err
+}
+
+func (r *byteArrayPageDecoder) ReadValues(values []Value) (int, error) {
+	i := 0
+	return r.readByteArrays(func(b []byte) (ok bool) {
+		if ok = i < len(values); ok {
+			values[i] = makeValueBytes(ByteArray, copyBytes(b))
+			values[i].columnIndex = r.columnIndex
+			i++
+		}
+		return ok
+	})
+}
+
+type fixedLenByteArrayPageDecoder struct {
+	decoder     encoding.Decoder
+	buffer      []byte
+	offset      int
+	remain      int
+	size        int
+	columnIndex int16
+}
+
+func newFixedLenByteArrayPageDecoder(typ Type, columnIndex int16, numValues int, decoder encoding.Decoder) *fixedLenByteArrayPageDecoder {
+	return &fixedLenByteArrayPageDecoder{
+		decoder:     decoder,
+		buffer:      getBuffer()[:],
+		remain:      numValues,
+		size:        typ.Length(),
+		columnIndex: ^columnIndex,
+	}
+}
+
+func (r *fixedLenByteArrayPageDecoder) Close() error {
+	putBuffer((*[bufferSize]byte)(r.buffer))
+	r.decoder = nil
+	r.buffer = nil
+	r.offset = 0
+	r.remain = 0
+	return nil
+}
+
+func (r *fixedLenByteArrayPageDecoder) ReadRequired(values []byte) (int, error) {
+	return r.ReadFixedLenByteArrays(values)
+}
+
+func (r *fixedLenByteArrayPageDecoder) ReadFixedLenByteArrays(values []byte) (n int, err error) {
+	if (len(values) % r.size) != 0 {
+		return 0, fmt.Errorf("cannot read FIXED_LEN_BYTE_ARRAY values of size %d into buffer of size %d", r.size, len(values))
+	}
+	if r.offset < len(r.buffer) {
+		i := copy(values, r.buffer[r.offset:])
+		n = i / r.size
+		r.offset += i
+		r.remain -= i
+		values = values[i:]
+	}
+	if r.decoder == nil {
+		err = io.EOF
+	} else {
+		var d int
+		values = values[:min(r.remain, len(values))]
+		d, err = r.decoder.DecodeFixedLenByteArray(r.size, values)
+		n += d
+		r.remain -= d
+		if r.remain == 0 && err == nil {
+			err = io.EOF
+		}
+	}
+	return n, err
+}
+
+func (r *fixedLenByteArrayPageDecoder) ReadValues(values []Value) (n int, err error) {
+	for {
+		for (r.offset+r.size) <= len(r.buffer) && n < len(values) {
+			values[n] = makeValueBytes(FixedLenByteArray, copyBytes(r.buffer[r.offset:r.offset+r.size]))
+			values[n].columnIndex = r.columnIndex
+			r.offset += r.size
+			r.remain -= r.size
+			n++
+		}
+
+		if r.remain == 0 || r.decoder == nil {
+			return n, io.EOF
+		}
+		if n == len(values) {
+			return n, nil
+		}
+
+		length := min(r.remain, cap(r.buffer))
+		buffer := r.buffer[:length]
+		d, err := r.decoder.DecodeFixedLenByteArray(r.size, buffer)
+		if d == 0 {
+			return n, err
+		}
+
+		r.buffer = buffer[:d*r.size]
+		r.offset = 0
+	}
+}
+
+var (
+	byteArrayLists sync.Pool // *encoding.ByteArrayList
```

Comment:
These comments are super helpful, thanks.

Reply:
Thanks! I took this habit because `sync.Pool` really says nothing about the objects that are expected to be stored in it, and the APIs are untyped so the compiler doesn't help either.

When we drop support for Go 1.17 we could put a generic API on `sync.Pool` and have `syncPool[*encoding.ByteArrayList]` which will be even better.